shared/src/main/scala/coop/rchain/dag/DagReaderSyntax.scala

```diff
@@ -0,0 +1,68 @@
+package coop.rchain.dag
+
+import cats.Show
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import coop.rchain.dag.DagOps.bfTraverseF
+import coop.rchain.shared.syntax._
+import fs2.Stream
+
+trait DagReaderSyntax {
+  implicit final def syntaxDagReader[F[_], A](
+      dag: DagReader[F, A]
+  ): DagReaderOps[F, A] = new DagReaderOps[F, A](dag)
+}
+
+final class DagReaderOps[F[_], A](
+    private val dag: DagReader[F, A]
+) extends AnyVal {
+
+  def parentsUnsafe(item: A)(
+      implicit sync: Sync[F],
+      line: sourcecode.Line,
+      file: sourcecode.File,
+      enclosing: sourcecode.Enclosing,
+      show: Show[A]
+  ): F[Seq[A]] = {
+    def source = s"${file.value}:${line.value} ${enclosing.value}"
+    def errMsg = s"Parents lookup failed: DAG is missing ${item.show}\n at $source"
+    dag.parents(item) >>= (_.liftTo(DagReader.VertexDoesNotExist(errMsg)))
+  }
+
+  def contains(item: A)(implicit sync: Sync[F]): F[Boolean] =
+    dag.parents(item).map(_.isDefined) ||^ dag.children(item).map(_.isDefined)
+
+  def mainParent(item: A)(implicit sync: Sync[F]): F[Option[A]] =
+    dag.parents(item).map(_.flatMap(_.headOption))
+
+  def mainParentUnsafe(item: A)(
+      implicit sync: Sync[F],
+      line: sourcecode.Line,
+      file: sourcecode.File,
+      enclosing: sourcecode.Enclosing,
+      show: Show[A]
+  ): F[A] = {
+    def source = s"${file.value}:${line.value} ${enclosing.value}"
+    def errMsg = s"Unsafe call for parents for block with no parents: ${item.show}\n at $source"
+    mainParent(item) >>= (_.liftTo(DagReader.VertexDoesNotExist(errMsg)))
+  }
+
+  def isInMainChain(
+      descendant: A,
+      ancestor: A,
+      height: A => F[Long]
```

Comment:
Help me to understand these arguments.

`descendant` is the target which the function want to check if the target is in the `MainChain`.
`ancestor` is the marking vertex which means the start poin of the `MainChain`.

Is that correct?

Reply:
For a dag like this
```
     b3
/         \
b1     b2
  \    /
   b0
```
blockHeightFunc for each blocks ->  `b0=1` , `b1=2`, `b2=2`, `b3=3`.

`isInMainChain(descendant=b1, ancestor=b3, height=blockHeightFunc)` what's the result of the function?

Reply:
The result will be false, as b1 is ancestor of b3, so function return false on first iteration of tail recursion
https://github.com/rchain/rchain/blob/fe40fe34335003babb42688618a6436f1d735d8e/shared/src/main/scala/coop/rchain/dag/DagReaderSyntax.scala#L74-L77


> Help me to understand these arguments.
> 
> `descendant` is the target which the function want to check if the target is in the `MainChain`.
> `ancestor` is the marking vertex which means the start poin of the `MainChain`.
> 
> Is that correct?

its other way. If you want to check if something is in the main chain - you start with some descendant (e.g. tip), as you want to traverse through main parents. There is no main child, you can traverse only upside down.
So `target` which we are going to evaluate agains `isInMainChain` predicate in caller function is **ancestor** here.